### ISSUE TITLE
why use torch.no_grad() int _get_indices() ??

### ISSUE THREAD

**Issue by kobenaxie** [facebookresearch/XLM#148]
https://github.com/facebookresearch/XLM/blob/46dd919b8ca1f22c11c4b121a6a8fd0d8fff7548/src/model/memory/memory.py#L589

I can understand, if use no_grad() here, does it mean that gradient will not backpropagation to `self.keys` ? and `self.keys` will never be updated ?

Anyone help ?

**Comment by glample** [comment]
This part of the code is only here to find the indices of the nearest neighbors. The scores are recomputed below given those indices: https://github.com/facebookresearch/XLM/blob/46dd919b8ca1f22c11c4b121a6a8fd0d8fff7548/src/model/memory/memory.py#L605-L610

This way, the keys are updated. However, I would recommend not using `HashingMemoryProduct`, but `HashingMemoryProductFast` instead, as it is faster both at training and inference (especially at training).

**Comment by kobenaxie** [comment]
OK, thank you for reply ~
